File: TP1/ex17_PalindromoRecursivoC/PalindromoRecursivo.c
/*
 *	Carlos Eduardo de Melo Sabino - 856290
 *	
 *	TP01Q17 - Palindromo Recursiovo em C
 */


#include <stdio.h>
#include <string.h>
#include <stdbool.h>


// Função recursiva para verificar se a string é um palindromo

bool ePalindromo(char str[], int i, int tamanho) {

	if ( i >= tamanho) {
        return true;
    }
    if (str[i] != str[tamanho]) {
        return false;
    }
    
	return ePalindromo(str, i + 1, tamanho - 1);
}

int main(void) {

	char str[1000];

	/*
	 * Não gosto de usar while e break mas foi o único
	 * jeito que encontrei e funcionou
	 */
	while (true) {

        scanf(" %[^\n]", str);

        if (strcmp(str, "FIM") == 0) {
            break;
        }
		else {
            
			int tamanho = strlen(str);
            
			if (ePalindromo(str, 0, tamanho - 1)) {
                printf("SIM\n");
            }
			else {
                printf("NAO\n");
            }
        }
    }
    return 0;
}

File: TP1/ex16_PalindromoRecursivoJava/PalindromoRecursivo.java
/*
 *		Carlos Eduardo de Melo Sabino - 856290
 *
 *		TP01Q16 - Palindromo Recursivo Java
 */

import java.util.*;

public class PalindromoRecursivo {

    // Função recursiva para verificar se a string é um palindromo
    public static boolean ePalindromo(String str, int i) { 
    
    	if (i >= str.length() / 2) {
        	return true;
    	}
         
    	/*
			Compara os caracteres do inicio até o meio e do
			final até o meio
		*/
    	if (str.charAt(i) != str.charAt(str.length() - i - 1)) {
        	 return false;
    	}
        
       // Chama a função recursivamente para comparar os proximos caracteres
       return ePalindromo(str, i + 1);
    } 

	public static void main(String[] args) {

        String str = "";

        Scanner sc = new Scanner(System.in);

        while (!(str = sc.nextLine()).equals("FIM")) {

            if (ePalindromo(str, 0)) {
                System.out.println("SIM");
            } 
			else {
                System.out.println("NAO");
            }

        }
    	sc.close();
    }
}	
